### PR TITLE
perf(provider/kubernetes): make one "large" caching agent

### PR DESCRIPTION
**Pull request by lwander**
It's worth pointing out that now it's more important to make use of the
"cacheThreads" flag in the account config simply because the default of
1 thread means at most one clouddriver can handle caching a single
account at any time.

Directly addresses: https://github.com/spinnaker/spinnaker/issues/2644
Is on the way to addressing: https://github.com/spinnaker/spinnaker/issues/2670

### CODE REVIEW

**Comment by lwander** [comment]
@wjoel FYI

**Comment by lwander** [comment]
This reduces the number of kubectl calls by a factor of roughly 25

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesCoreCachingAgent.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    types.add(AUTHORITATIVE.forType(Keys.LogicalKind.CLUSTERS.toString()));
+
+    types.addAll(primaryKinds().stream().map(k -> AUTHORITATIVE.forType(k.toString())).collect(Collectors.toList()));
+
+    return Collections.unmodifiableSet(new HashSet<>(types));
+  }
+
+  @Override
+  protected List<KubernetesKind> primaryKinds() {
+    return Arrays.asList(
```
Comment (danielpeach) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesCoreCachingAgent.java]: Ideally this list would be generated such that any dynamically registered `KubernetesKind` would be included here as well.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesCoreCachingAgent.java [hunk 2]
```diff
@@ -45,20 +47,23 @@
...
+    types.addAll(primaryKinds().stream().map(k -> AUTHORITATIVE.forType(k.toString())).collect(Collectors.toList()));
+
+    return Collections.unmodifiableSet(new HashSet<>(types));
   }
 
   @Override
-  protected KubernetesKind primaryKind() {
-    return KubernetesKind.ROLE;
+  protected List<KubernetesKind> primaryKinds() {
+    synchronized (KubernetesKind.getValues()) {
```
Comment (jtk54) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesCoreCachingAgent.java]: Is there a reason this `synchronized` isn't scoped to the function?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/handler/KubernetesStatefulSetHandler.java [hunk 3]
```diff
@@ -138,4 +146,29 @@ private Status status(V1beta2StatefulSet statefulSet) {
 
     return result;
   }
+
+  @Override
+  public void addRelationships(Map<KubernetesKind, List<KubernetesManifest>> allResources, Map<KubernetesManifest, List<KubernetesManifest>> relationshipMap) {
+    BiFunction<String, String, String> manifestName = (namespace, name) -> namespace + ":" + name;
```
Comment (jtk54) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/handler/KubernetesStatefulSetHandler.java]: Used more than once or twice, might warrant its own util function.

**Comment by jtk54** [comment]
Welp.

**Pull request merged**